### PR TITLE
SALTO-4968: Cache Salesforce Client listMetadataObjects results

### PR DESCRIPTION
**Pull request by tamtamirr**
This PR adds caching on the results of **listMetadataObjects**.

---

Main reason for this caching is the fact we have some MetadataTypes we list more than once, especially in **fetch with changes detection mode**. Listing types can be time consuming, depending on the amount of **Instances** per type. It's widely noticeable around **CustomFields** and **CustomObjects**.

---
_Release Notes_: 
_None_

---
_User Notifications_: 
_None_

### CODE REVIEW

File: packages/salesforce-adapter/src/client/client.ts [hunk 1]
```diff
@@ -635,12 +637,38 @@ export default class SalesforceClient {
...
+    }
+    const [cachedQueries, nonCachedQueries] = _.partition(
+      makeArray(listMetadataQuery),
+      query => Object.keys(this.filePropsByType).includes(query.type)
+    )
+    const cachedProps = cachedQueries.flatMap(query => makeArray(this.filePropsByType[query.type]))
+    if (nonCachedQueries.length === 0) {
+      log.debug('returning cached listMetadataObjects for %s', safeJsonStringify(listMetadataQuery))
+      return { result: cachedProps, errors: [] }
+    }
```
Comment (tomsellek) [inline packages/salesforce-adapter/src/client/client.ts:664]: If you taught `sendChunkedList` to handle empty arrays as its parameter this `if` would not be necessary, right?

**Review by tomsellek — APPROVED** [review]
LGTM. 1 question on a possible simplification.

File: packages/salesforce-adapter/src/client/client.ts (hunk 1 above)
Comment (tamtamirr) [reply packages/salesforce-adapter/src/client/client.ts:664]: This if is relevant when all of the queried typed were already queried before. So we return them from the cache :)

Comment (tamtamirr) [reply packages/salesforce-adapter/src/client/client.ts:664]: It just makes more sense IMO to execute this `if` before invoking `sendChunkedList`.  L661-664 will have to be either here or within the private method **sendChunkedList** so I don't see how it simplifies the code TBH 🤔

**Pull request merged**